In [ ]:
# 코랩을 시작할 때 아래코드를 한 번 돌려줍니다.
!pip install selenium
!apt-get update
!apt install chromium-chromedriver
!cp /usr/lib/chromium-browser/chromedriver /usr/bin

Hit:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease
Hit:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease
Hit:3 http://archive.ubuntu.com/ubuntu jammy InRelease
Hit:4 http://security.ubuntu.com/ubuntu jammy-security InRelease
Hit:5 http://archive.ubuntu.com/ubuntu jammy-updates InRelease
Hit:6 http://archive.ubuntu.com/ubuntu jammy-backports InRelease
Hit:7 https://ppa.launchpadcontent.net/c2d4u.team/c2d4u4.0+/ubuntu jammy InRelease
Hit:8 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease
Hit:9 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease
Hit:10 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease
Reading package lists... Done
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
chromium-chromedriver is already the newest version (1:85.0.4183.83-0ubuntu2.22.04.1).
0 upgraded, 0 newly installed, 0 to remove and 38 n

3) 사이트 호출

직접 들어가서 어떤 화면을 열지 확인하고 그 URL를 가지고 온다.

다음과 같이 실행하면 아나콘다, 파이참등에서는 크롬 브라우저가 열리지만, 코랩에서는 크롬브라우저가 별도로 열리지 않는다. 그냥 가상공간에 떠 있다고 상상하면 된다.

In [ ]:
import requests
from bs4 import BeautifulSoup
from selenium import webdriver
from time import sleep
import pandas as pd
import time
from selenium.webdriver.common.by import By

https://blog.naver.com/j7youngh/222852618427
셀레니움 설명 친절하시다

In [ ]:
options = webdriver.ChromeOptions()
options.add_argument('--headless')  # 브라우저를 표시하지 않고 실행하려면 이 옵션을 추가합니다.
options.add_argument('--no-sandbox')
options.add_argument('--disable-dev-shm-usage')

# 코랩에서 크롬 드라이버 실행
driver =  webdriver.Chrome(options=options)

URL = 'https://page.kakao.com/menu/10010/screen/52'

# 웹 드라이버 초기화
# driver = webdriver.Chrome()

driver.get(URL)
time.sleep(2)

# 사이드 탭 클릭 후 웹툰 링크 가져오기
# 다음으로 넘어가는 법: 패턴을 찾기보다는 다음 버튼을 누르면 다음 페이지로 넘어가기 때문에
# 버튼의 경로나 요소를 찾아 click()함수를 사용하자


# 신작, 요일 별, 완결 카테고리 링크 추출
category_list = []

# 첫 번째 div 태그 중에서 class가 "flex justify-center overflow-hidden"인 것을 찾기
div_elements = driver.find_elements(By.CSS_SELECTOR, 'div.flex.justify-center.overflow-hidden')

# 첫 번째 반복문
for div_element in div_elements:
    # div 태그의 자식인 a 태그 찾기
    a_tag = div_element.find_element(By.TAG_NAME, 'a')

    # a 태그의 href 속성 가져오기
    href = a_tag.get_attribute('href')
    category_list.append(href)
    print(href)


# 각 링크를 들어간 후 해당 웹툰 링크 들어감 -> 제목, 세부 정보, 태크 추출

# 웹 드라이버 종료
driver.quit()

https://page.kakao.com/menu/10010/screen/52?tab_uid=11
https://page.kakao.com/menu/10010/screen/52?tab_uid=1
https://page.kakao.com/menu/10010/screen/52?tab_uid=2
https://page.kakao.com/menu/10010/screen/52?tab_uid=3
https://page.kakao.com/menu/10010/screen/52?tab_uid=4
https://page.kakao.com/menu/10010/screen/52?tab_uid=5
https://page.kakao.com/menu/10010/screen/52?tab_uid=6
https://page.kakao.com/menu/10010/screen/52?tab_uid=7
https://page.kakao.com/menu/10010/screen/52?tab_uid=12


In [ ]:
# @title 카테고리 안의 웹툰 링크 추출
driver =  webdriver.Chrome(options=options)

url = category_list[0]
driver.get(url)

webtoon_list = []

# 'div.w-full.overflow-hidden' 내부에 있는 모든 'a' 태그 선택
a_tags = driver.find_elements(By.CSS_SELECTOR, 'div.w-full.overflow-hidden div div div a')

for a_tag in a_tags:
    # 각 'a' 태그의 href 속성(링크) 출력
    print(a_tag.get_attribute('href'))
    # webtoon_list.append(a_tag.get_attribute('href'))

driver.quit()

https://page.kakao.com/content/63665097
https://page.kakao.com/content/63655107
https://page.kakao.com/content/63612419
https://page.kakao.com/content/63646168
https://page.kakao.com/content/63663567
https://page.kakao.com/content/63555983
https://page.kakao.com/content/63604804
https://page.kakao.com/content/63492925
https://page.kakao.com/content/63500287
https://page.kakao.com/content/63555003
https://page.kakao.com/content/63529460
https://page.kakao.com/content/62074243
https://page.kakao.com/content/63512854
https://page.kakao.com/content/63604826
https://page.kakao.com/content/63672731
https://page.kakao.com/content/63733643
https://page.kakao.com/content/63674828
https://page.kakao.com/content/63572423
https://page.kakao.com/content/63529403
https://page.kakao.com/content/63539722
https://page.kakao.com/content/63154945
https://page.kakao.com/content/63364797
https://page.kakao.com/content/63497244
https://page.kakao.com/content/62932582
https://page.kakao.com/content/63672738


In [ ]:
# @title 스크롤이용해서 전체 링크 추출
driver = webdriver.Chrome(options=options)

url = category_list[0]
driver.get(url)

webtoon_list = []
last_height = driver.execute_script("return document.body.scrollHeight")

while True:
  # 페이지 끝까지 스크롤 다운
  driver.execute_script("window.scrollTo(0, document.body.scrollHeight);")

  # 페이지 로드를 기다림
  time.sleep(1)

  # 새로운 스크롤 높이 계산 및 이전 스크롤 높이와 비교
  new_height = driver.execute_script("return document.body.scrollHeight")
  if new_height == last_height:
    break
  last_height = new_height

  # 'div.w-full.overflow-hidden' 내부에 있는 모든 'a' 태그 선택
  a_tags = driver.find_elements(By.CSS_SELECTOR, 'div.w-full.overflow-hidden div div div a')

  for a_tag in a_tags:
      # webtoon_list.append(a_tag.get_attribute('href'))
      print(a_tag.get_attribute('href'))

driver.quit()

https://page.kakao.com/content/63665097
https://page.kakao.com/content/63655107
https://page.kakao.com/content/63612419
https://page.kakao.com/content/63646168
https://page.kakao.com/content/63663567
https://page.kakao.com/content/63555983
https://page.kakao.com/content/63604804
https://page.kakao.com/content/63492925
https://page.kakao.com/content/63500287
https://page.kakao.com/content/63555003
https://page.kakao.com/content/63529460
https://page.kakao.com/content/62074243
https://page.kakao.com/content/63512854
https://page.kakao.com/content/63604826
https://page.kakao.com/content/63672731
https://page.kakao.com/content/63733643
https://page.kakao.com/content/63674828
https://page.kakao.com/content/63572423
https://page.kakao.com/content/63529403
https://page.kakao.com/content/63539722
https://page.kakao.com/content/63154945
https://page.kakao.com/content/63364797
https://page.kakao.com/content/63497244
https://page.kakao.com/content/62932582
https://page.kakao.com/content/63672738


# 최종 코드

In [ ]:
def webData(input_link):
    driver = webdriver.Chrome(options=options)

    url = input_link
    driver.get(url)

    webtoon_list = []
    last_height = driver.execute_script("return document.body.scrollHeight")

    while True:
        # 페이지 끝까지 스크롤 다운
        driver.execute_script("window.scrollTo(0, document.body.scrollHeight);")

        # 페이지 로드를 기다림
        time.sleep(1)

        # 새로운 스크롤 높이 계산 및 이전 스크롤 높이와 비교
        new_height = driver.execute_script("return document.body.scrollHeight")
        if new_height == last_height:
            break
        last_height = new_height

    # 'div.w-full.overflow-hidden' 내부에 있는 모든 'a' 태그 선택
    a_tags = driver.find_elements(By.CSS_SELECTOR, 'div.w-full.overflow-hidden div div div a')

    for a_tag in a_tags:
        webtoon_list.append(a_tag.get_attribute('href'))

    driver.quit()
    return webtoon_list


In [ ]:
for i in category_list:
  q = webData(i)

print(q)

['https://page.kakao.com/content/50866481', 'https://page.kakao.com/content/63492925', 'https://page.kakao.com/content/60564669', 'https://page.kakao.com/content/53190884', 'https://page.kakao.com/content/61579814', 'https://page.kakao.com/content/55166366', 'https://page.kakao.com/content/54526433', 'https://page.kakao.com/content/63674828', 'https://page.kakao.com/content/62011604', 'https://page.kakao.com/content/63539722', 'https://page.kakao.com/content/62477346', 'https://page.kakao.com/content/63348499', 'https://page.kakao.com/content/55249956', 'https://page.kakao.com/content/51415682', 'https://page.kakao.com/content/54439672', 'https://page.kakao.com/content/62830615', 'https://page.kakao.com/content/60636581', 'https://page.kakao.com/content/63154945', 'https://page.kakao.com/content/62346577', 'https://page.kakao.com/content/52226628', 'https://page.kakao.com/content/60636644', 'https://page.kakao.com/content/54957639', 'https://page.kakao.com/content/62075272', 'https://p

In [ ]:
print(len(q))

1868


In [ ]:
driver = webdriver.Chrome(options=options)

url = q[0]
driver.get(url)

a_tags = driver.find_elements(By.CSS_SELECTOR, '.relative.flex.items-center.justify-center a')

if len(a_tags) > 1:
    second_link = a_tags[1]
    second_link.click()

    time.sleep(1)
    # 원하는 요소를 CSS 선택자로 찾기
    title_element = driver.find_element(By.CSS_SELECTOR, 'span.font-large3-bold.mb-3pxr.text-ellipsis.break-all.text-el-70.line-clamp-2')
    summary_element = driver.find_element(By.CSS_SELECTOR, 'span.font-small1.mb-8pxr.block.whitespace-pre-wrap.break-words.text-el-70')
    # div 태그 내에서 a 태그를 찾아서 data-t-obj 속성 값을 가져오기
    category_tag_elements = driver.find_elements(By.CSS_SELECTOR, 'div.flex.w-full.flex-wrap.px-18pxr.pb-10pxr a.mb-8pxr.mr-8pxr')

    # 데이터 프레임 생성을 위한 빈 리스트 초기화
    titles = []
    summaries = []
    category_tags = []

    # 요소에서 텍스트 추출 및 출력
    if title_element:
        text_title = title_element.text
        text_summary = summary_element.text
        text_category_tag = []

        for element in category_tag_elements:
          data_t_obj = element.get_attribute('data-t-obj')

          if 'click' in data_t_obj:
            copy_info = data_t_obj.split('"copy":')[1].split('}')[0]
            text_category_tag.append(copy_info.strip('"'))
        category_tags_combined = ', '.join(text_category_tag)

        print("찾은 요소의 텍스트:", text_title)
        print("찾은 요소의 텍스트:", text_summary)
        print("찾은 요소의 텍스트:", category_tags_combined)

        webDataFrame = pd.DataFrame({
            'Title': [text_title],
            'Summary': [text_summary],
            'Category_Tag': [category_tags_combined]
        })
    else:
        print("원하는 요소를 찾을 수 없습니다.")


else:
    print("두 번째 링크가 존재하지 않습니다.")


driver.quit()


찾은 요소의 텍스트: 나 혼자만 레벨업
찾은 요소의 텍스트: 10여 년 전, 
다른 차원과 이쪽 세계를 이어 주는 통로 ‘게이트’가 열리고 
평범한 이들 중 각성한 자들이 생겨났다.

게이트 안의 던전에서 마물을 사냥하는 각성자.
그들을 일컬어 ‘헌터’라 부른다.

그러나 모든 헌터가 강한 것은 아니다.

내 이름은 성진우.
E급 헌터다. 

저급 던전에서조차 죽을 고비를 넘겨야 하는
‘인류 최약병기’. 

무엇 하나 내세울 것 없는 형편에
저급 던전으로 근근이 생계를 이어가던 그는,
D급 던전 속에 숨겨진 
최악의 난이도의 이중 던전을 만난다.

그리고 결국… 죽음을 맞이하려던 순간,
그는 기이한 능력을 얻게 된다. 

오직 그에게만 보이는 게임 퀘스트 창!
오직 그만이 알게 된 레벨 업의 비밀!

퀘스트를 따라 수련하고
몬스터를 사냥하면 레벨이 오른다.
오직 그 혼자만! 


최약체 헌터에서 최강 헌터로 각성하다!

추공 작가의 인기 판무 소설 <나 혼자만 레벨업>이
웹툰으로 드디어 등장!

************************************************
[표지 및 로고 디자인] 디자인그룹 헌드레드

************************************************
웹툰 [나 혼자만 레벨업]은 매주 목요일, 주1회 연재됩니다.
찾은 요소의 텍스트: 생존, 먼치킨, 시스템, 천재, 노력, 성장, 이능력, 플레이어


In [ ]:
webDataFrame

Title                                            Summary  \
0  나 혼자만 레벨업  10여 년 전, \n다른 차원과 이쪽 세계를 이어 주는 통로 ‘게이트’가 열리고 \...   

                          Category_Tag  
0  생존, 먼치킨, 시스템, 천재, 노력, 성장, 이능력, 플레이어

In [ ]:
from selenium.common.exceptions import NoSuchElementException

In [ ]:
import pandas as pd
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.common.exceptions import NoSuchElementException
import time

options = webdriver.ChromeOptions()
options.add_argument('--headless')  # 브라우저를 표시하지 않고 실행하려면 이 옵션을 추가합니다.
options.add_argument('--no-sandbox')
options.add_argument('--disable-dev-shm-usage')

# 코랩에서 크롬 드라이버 실행
driver =  webdriver.Chrome(options=options)

def webtoonData(inputWebtoonLink):
    driver = webdriver.Chrome(options=options)

    url = inputWebtoonLink
    driver.get(url)

    a_tags = driver.find_elements(By.CSS_SELECTOR, '.relative.flex.items-center.justify-center a')
    not_link = []  # not_link 변수 초기화

    if len(a_tags) > 1:
        second_link = a_tags[1]
        second_link.click()

        time.sleep(1)

        try:
            # 원하는 요소를 CSS 선택자로 찾기
            title_element = driver.find_element(By.CSS_SELECTOR, 'span.font-large3-bold.mb-3pxr.text-ellipsis.break-all.text-el-70.line-clamp-2')
            summary_element = driver.find_element(By.CSS_SELECTOR, 'span.font-small1.mb-8pxr.block.whitespace-pre-wrap.break-words.text-el-70')
            # div 태그 내에서 a 태그를 찾아서 data-t-obj 속성 값을 가져오기
            category_tag_elements = driver.find_elements(By.CSS_SELECTOR, 'div.flex.w-full.flex-wrap.px-18pxr.pb-10pxr a.mb-8pxr.mr-8pxr')

            # 데이터 프레임 생성을 위한 빈 리스트 초기화
            titles = []
            summaries = []
            category_tags = []


            # 요소에서 텍스트 추출 및 리스트에 추가
            if title_element:
                text_title = title_element.text
                text_summary = summary_element.text
                text_category_tag = []

                for element in category_tag_elements:
                    data_t_obj = element.get_attribute('data-t-obj')

                    if 'click' in data_t_obj:
                        copy_info = data_t_obj.split('"copy":')[1].split('}')[0]
                        text_category_tag.append(copy_info.strip('"'))

                category_tags_combined = ', '.join(text_category_tag)
                text_summary = text_summary.replace('\n', "")

                # 리스트에 추가
                titles.append(text_title)
                summaries.append(text_summary)
                category_tags.append(category_tags_combined)

            else:
                # print(f"두 번째 링크에 대한 정보가 존재하지 않습니다: {url}")
                driver.quit()
                return None, [url], 1  # 수정: not_link을 리스트로 반환하고, current_count도 함께 반환

        except NoSuchElementException as e:
            # print(f"두 번째 링크에 대한 정보를 찾을 수 없습니다: {e}")
            driver.quit()
            return None, [url], 1  # 수정: not_link을 리스트로 반환하고, current_count도 함께 반환

    else:
        # print(f"두 번째 링크가 존재하지 않습니다: {url}")
        not_link.append(url)
        driver.quit()
        return None, not_link, 1  # 수정: not_link을 리스트로 반환하고, current_count도 함께 반환

    driver.quit()

    # 데이터 프레임 생성
    df = pd.DataFrame({
        'Title': titles,
        'Summary': summaries,
        'Category_Tag': category_tags
    })

    return df, not_link, 1  # 수정: not_link을 리스트로 반환하고, current_count도 함께 반환

# 예시로 200번 수집할 때까지 호출
category_list = q[1300:]
result_dfs = []
result_not_links = []

for index, link in enumerate(category_list, start=1):
    result_df, not_links, current_count = webtoonData(link)

    if result_df is not None:
        result_dfs.append(result_df)
    else:
        result_not_links.extend([(link, count) for count in not_links])

    # 100번마다 현재 수집 정도 출력
    if index % 5 == 0:
        print(f"현재 수집 정도: {index}")

# 최종 결과 데이터프레임 생성
final_df = pd.concat(result_dfs, ignore_index=True)

print(f"최종 수집 완료: {len(final_df)}")


현재 수집 정도: 5
현재 수집 정도: 10
현재 수집 정도: 15
현재 수집 정도: 20
현재 수집 정도: 25
현재 수집 정도: 30
현재 수집 정도: 35
현재 수집 정도: 40
현재 수집 정도: 45
현재 수집 정도: 50
현재 수집 정도: 55
현재 수집 정도: 60
현재 수집 정도: 65
현재 수집 정도: 70
현재 수집 정도: 75
현재 수집 정도: 80
현재 수집 정도: 85
현재 수집 정도: 90
현재 수집 정도: 95
현재 수집 정도: 100
현재 수집 정도: 105
현재 수집 정도: 110
현재 수집 정도: 115
현재 수집 정도: 120
현재 수집 정도: 125
현재 수집 정도: 130
현재 수집 정도: 135
현재 수집 정도: 140
현재 수집 정도: 145
현재 수집 정도: 150
현재 수집 정도: 155
현재 수집 정도: 160
현재 수집 정도: 165
현재 수집 정도: 170
현재 수집 정도: 175
현재 수집 정도: 180
현재 수집 정도: 185
현재 수집 정도: 190
현재 수집 정도: 195
현재 수집 정도: 200
현재 수집 정도: 205
현재 수집 정도: 210
현재 수집 정도: 215
현재 수집 정도: 220
현재 수집 정도: 225
현재 수집 정도: 230
현재 수집 정도: 235
현재 수집 정도: 240
현재 수집 정도: 245
현재 수집 정도: 250
현재 수집 정도: 255
현재 수집 정도: 260
현재 수집 정도: 265
현재 수집 정도: 270
현재 수집 정도: 275
현재 수집 정도: 280
현재 수집 정도: 285
현재 수집 정도: 290
현재 수집 정도: 295
현재 수집 정도: 300
현재 수집 정도: 305
현재 수집 정도: 310
현재 수집 정도: 315
현재 수집 정도: 320
현재 수집 정도: 325
현재 수집 정도: 330
현재 수집 정도: 335
현재 수집 정도: 340
현재 수집 정도: 345
현재 수집 정도: 350
현재 수집 정도: 355
현재 수집 정도: 360
현재 수집 정도: 36

In [ ]:
final_df

Title  \
0                  그는 내 여동생을 사랑했다   
1                          로맨틱 피치   
2    메리지 B ~회귀한 영애는 가문의 원수와 결혼한다~   
3                            일월성신   
4                     네 이웃에게 친절하라   
..                            ...   
186              왕가의 후예 [고화질, 완결]   
187                            마왕   
188                          매의 검   
189                        신수의 주인   
190               신의 선물 [고화질, 완결]   

                                               Summary  \
0    내가 좋아하는 그 남자는 내 여동생을 사랑한다?! 어긋나고 어긋나는 사랑 이야기!"...   
1    오래된 신문 보급소 딸로서 신문과 함께 자라왔기에 신문기자의 꿈을 꾸게 된 혜진. ...   
2    "네가 멜리시를 죽인 거야!네가 나를 놓아줬더라면, 그녀는 살아 있을 텐데!"사랑하...   
3    태초의 성신(星神)은 천상의 신들이 본분조차 잊고 탐욕에 휩싸이자오직 '신탁의 인간...   
4    '점장님도 철 좀 들었으면, 할머니를 모시고 사는 유튜버 '함5함5'처럼.'한 시골...   
..                                                 ...   
186  라·엘 파스타나 이집트군에 의해 일족을 몰살당한 아메스의 제일왕녀!아메스의 재건과 ...   
187  [매주 금요일 업데이트 됩니다!] 평범한 일상을 보내던 고등학생 장세나. 어느 날 ...   
188  ※ 1화부터 5화까지는 흑백이며, 6화부터 컬러입니다.  인기 로맨스소설가 김경미의...   
189            인간소녀, 드래곤 그리고 쌍둥이 오빠!? 그대가 원하는 판타지 로맨스!   
190  “해와 달이 만나 정을 이루고 어둠 속에서 빛이 드러나 온조의 정수리를 치리라.”예...   

                                          Category_Tag  
0    로맨스판타지, 오해물, 착각물, 영혼체인지, 짝사랑, 첫사랑, 소유욕/독점욕, 왕족...  
1               현대로맨스, 캠퍼스물, 우연한만남, 까칠남, 엉뚱발랄녀, 로맨틱코미디  
2    로맨스판타지, 타임슬립, 성장물, 환생물, 회귀물, 중세물, 여주판타지, 운명, 나...  
3    판타지물, 현대물, 힐링물, 성장물, 신령, 전생/환생, 초월적존재, 환생물, 가족...  
4         미스터리, 스릴러, 현대물, 기묘한, 남주중심, 무서운, 두근거리는, 이야기중심  
..                                                 ...  
186                                                     
187                                                     
188                                                     
189                                로맨스판타지, 역하렘, 로맨틱코미디  
190                                                     

[191 rows x 3 columns]

In [ ]:
import os
import json
import pandas as pd

# JSON 파일들이 있는 디렉토리 경로
json_directory = 'webtoon/'

# 디렉토리 내의 모든 JSON 파일 읽기
json_files = [f for f in os.listdir(json_directory) if f.endswith('.json')]

# 데이터프레임 초기화할 빈 리스트
df_list = []

# 각 JSON 파일에 대해 데이터프레임으로 변환하여 리스트에 추가
for json_file in json_files:
    with open(os.path.join(json_directory, json_file), 'r') as file:
        json_data = json.load(file)
        df_list.append(pd.json_normalize(json_data))

# 모든 데이터프레임을 합치기
final_df = pd.concat(df_list, ignore_index=True)


          id    type                title               genre  year theme  \
0    04_3810  comics            우당탕 대모험 9          [판타지, 드라마]  2010    모험   
1    04_3874  comics          새콤달콤 청춘예찬 3       [멜로/로맨스, 드라마]  2021    사랑   
2    04_3367  comics         쌍검 마법사 대활극 5           [판타지, 액션]  2012    모험   
3    04_3879  comics          새콤달콤 청춘예찬 8       [멜로/로맨스, 드라마]  2021    사랑   
4    04_3532  comics  마을의 해결사가 고교생이라고? 14      [액션, 드라마, 판타지]  2017    도전   
..       ...     ...                  ...                 ...   ...   ...   
363  04_2506  comics           친해질 수 있을까?          [판타지, 드라마]  2018    모험   
364  04_3505  comics        소년은 가문을 구한다 3           [액션, 판타지]  2017    도전   
365  04_3279  comics     저, 가입한 적 없는데요? 5       [멜로/로맨스, 드라마]  2004    사랑   
366  04_2819  comics         눈 떠 보니 이세계 2  [드라마, 멜로/로맨스, 판타지]  2009    모험   
367  04_3415  comics         나와 당신과 그녀 10          [드라마, 코미디]  2018    사랑   

                                               concept   structure      mot

In [ ]:
webtoon2_data = final_df[['title','main_character','motif','genre','theme','conflict']]
webtoon2_data.head(3)
webtoon2_data.to_csv('webtoon2_data.csv', index=False)

In [ ]:
import os
import json
import pandas as pd

# JSON 파일들이 있는 디렉토리 경로
json_directory = 'novel/'

# 디렉토리 내의 모든 JSON 파일 읽기
json_files = [f for f in os.listdir(json_directory) if f.endswith('.json')]

# 데이터프레임 초기화할 빈 리스트
df_list = []

# 각 JSON 파일에 대해 데이터프레임으로 변환하여 리스트에 추가
for json_file in json_files:
    with open(os.path.join(json_directory, json_file), 'r') as file:
        json_data = json.load(file)
        df_list.append(pd.json_normalize(json_data))

# 모든 데이터프레임을 합치기
novel_df = pd.concat(df_list, ignore_index=True)

In [ ]:
novel_data = novel_df[['title','main_character','motif','genre','theme','conflict']]
novel_data.head(3)
novel_data.to_csv('novel_data.csv', index=False)

In [ ]:
# 전처리 및 합병
webtoon2_df = pd.read_csv('webtoon2_data.csv')
novel_df = pd.read_csv('novel_data.csv')

web_noval_df = pd.concat([webtoon2_df, pd.DataFrame(novel_df, columns=['title','main_character','motif','genre','theme','conflict'])], ignore_index=True)
print(len(web_noval_df))

536


In [ ]:
web_noval_df['Category_Tag'] = web_noval_df[['main_character', 'motif', 'genre', 'theme']].apply(
    lambda x: ', '.join([str(e).replace(' ', '').replace("'", "").replace(",",", ").strip('[]') for e in x]), axis=1)
web_noval_df = web_noval_df.drop(['main_character', 'motif', 'genre', 'theme'], axis=1)

web_noval_df = web_noval_df.rename(columns={'title': 'Title', 'conflict': 'Summary'})

web_noval_df.head(3)

Title                                            Summary  \
0     우당탕 대모험 9  C002는 C001과 C003이 가깝게 지내자, 자신을 따돌리고 있다는 생각에 소외...   
1   새콤달콤 청춘예찬 3          C002는 첫사랑 상대 C001와는 잘 지내지만 친구를 사귀기 어려워한다.   
2  쌍검 마법사 대활극 5  C001는 마녀 C016와의 전투에서 이기기 위해 궁극기1을 사용한 결과, 극심한 ...   

                   Category_Tag  
0    소년영웅, 오해(착각), 판타지, 드라마, 모험  
1  평범, 첫눈에반한사랑, 멜로/로맨스, 드라마, 사랑  
2      헌터, 마법사(마녀), 판타지, 액션, 모험

In [ ]:
not_link_data3 = result_not_links
print(len(result_not_links))

377


In [ ]:
# CSV 파일로 저장
final_df.to_csv('webtoon_data5.csv', index=False)

In [ ]:
# 총 수집된 데이터 하나로 합치기

df1 = pd.read_csv('webtoon_data.csv')
df2 = pd.read_csv('webtoon_data2.csv')
df3 = pd.read_csv('webtoon_data3.csv')
df4 = pd.read_csv('webtoon_data4.csv')
df5 = pd.read_csv('webtoon_data5.csv')


total_df = pd.concat([df1, df2, df3, df4, df5], ignore_index=True)
total_df.to_csv('total_webtoon_data.csv', index=False)

In [ ]:
total_df = pd.concat([total_df, web_noval_df], ignore_index=True)
print(len(total_df))
total_df.to_csv('total_webtoon_data.csv', index=False)

1023


In [ ]:
!pip install python-mecab-ko

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 573.9/573.9 kB 6.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 34.5/34.5 MB 10.5 MB/s eta 0:00:00


# 데이터 전처리 진행

In [ ]:
import pandas as pd
import numpy as np

from datetime import datetime
import time
import re

from mecab import MeCab

In [ ]:
# 데이터 불러오기
df = pd.read_csv(file_path + 'total_webtoon_data(2).csv')

In [ ]:
df.isnull().sum()

Title             0
Summary           0
Category_Tag    179
dtype: int64

In [ ]:
missing_category_rows = df[df['Category_Tag'].isnull()]
missing_category_rows

Title                                            Summary  \
8                  무빙  남들이 모르는 비밀을 간직하고 있는 고등학생. 그리고 그 부모들의 이야기. 소년소녀...   
18             메디컬 환생  실패한 외과 의사로서의 삶을 살던 김진현,우연한 사건으로 인해 환생의 기회를얻는다....   
20               블랙윈터  조선 후기, 백송골이라는 시골 마을에 사는 13세 소녀 설해사.친척과의 갈등으로 오...   
29          식사하고 가세요!  미각을 잃은 요리사 이림이 좌절에 빠진 와중 가상현실게임인 [리얼]에서 오감을 체험...   
32               우렁강도  강도와의 유쾌하고 살벌한 동거 라이프!  지호는 한순간의 나태로 인해 기숙사 신청 ...   
..                ...                                                ...   
480              프라이드  ‘스타’.매순간 높은 곳에서 아름답게 빛나며그저 바라보는 것만으로도 행복하게 만드는...   
482  왕가의 후예 [고화질, 완결]  라·엘 파스타나 이집트군에 의해 일족을 몰살당한 아메스의 제일왕녀!아메스의 재건과 ...   
483                마왕  [매주 금요일 업데이트 됩니다!] 평범한 일상을 보내던 고등학생 장세나. 어느 날 ...   
484              매의 검  ※ 1화부터 5화까지는 흑백이며, 6화부터 컬러입니다.  인기 로맨스소설가 김경미의...   
486   신의 선물 [고화질, 완결]  “해와 달이 만나 정을 이루고 어둠 속에서 빛이 드러나 온조의 정수리를 치리라.”예...   

    Category_Tag  
8            NaN  
18           NaN  
20           NaN  
29           NaN  
32           NaN  
..           ...  
480          NaN  
482          NaN  
483          NaN  
484          NaN  
486          NaN  

[179 rows x 3 columns]

179.. 전처리 어케하지
<br> 기존 카테고리에서 줄거리 기반 장르 머신러닝 분석 진행.
<br> 값이 없는 걸 test 데이터로 놓고 출력 값을 카테고리로 저장

### 다중 클래스 분류 : 머신러닝 사용



In [ ]:
import copy

# DataFrame 복사
new_df = copy.deepcopy(df)

not_nan_df = new_df[pd.notna(new_df['Category_Tag'])]
nan_df = new_df[pd.isna(new_df['Category_Tag'])]

print(len(not_nan_df), len(nan_df))

1380 179


In [ ]:
# 카테고리 라벨링
label = []
label_dict = {}

for i in new_df['Category_Tag']:
    if isinstance(i, str):
        text = ''.join(i)
        text = text.split(' ')
        label.extend(text)
    else:
        pass

label = list(set(label))
print(len(label))
print(label)

for index, labeles in enumerate(label):
    label_dict[labeles.strip()] = str(index)

print(label_dict)

682
['라이벌,', '이종족,', '순진함,', '이능력,', '빙의,', '사랑', '로맨틱코미디', '빙의물,', '치유물,', '대형견남,', '금단의관계,', '잔잔한', '감춰진혈통찾기,', '마검사,', '공포(호러),', '성경험이없음,', '신,', '엉뚱함,', '사제지간', '상처남', '검,', '귀염,', '판타지,SF,미스터리,', '뇌섹녀,', '오해,', '뜻밖의편지,', '능력남', '모험', '상승', '학원물,', '희생', '갑작스런사고,', '조력자,', '집착공,', '엉뚱발랄녀', '이계방문,', '개그물', '남장여자,', '돌연변이(뮤턴트),', '남주중심,', '배반자,', '영유아,', '조신남,', '발견', '시스템,', '출세주의자,', '힐링물,', '라이벌', '무심,', '예술가,', '악녀', '영혼체인지,', '도제수련,', '아포칼립스,', '다정,', '메디컬물,', '변장,', '회귀/타임슬립,', '의적,', '액션,SF,', '영혼체인지/빙의,', '나이차를극복한사랑,', '동거물,', '다정녀,', '로맨틱코미디,', '신화물,', '개그물,', '무림', '계략적,', '시한부,', '판타지,SF,액션,', '도망,', '여주중심', '드라마,판타지,멜로/로맨스,', '재벌녀,', '소림,', '소꿉친구,', '회사원', '고귀한창녀,', '고등사기꾼,', '스포츠물,', '미스터리,스릴러,', '정파,', '학생,', '음흉함,', '여주판타지,', '판타지,코미디,드라마,', '도박,', '도사,', '중세물,', '신분차이,', '전생/환생,', '군인,', '계략녀', '판타지,드라마,', '공황장애,', '희생하는부모,', '형사(수사관),', '집착,', '오피스물', '작가,', '귀향,', '도망자,', '지구종말,', '냉정,', '신분위장,', '수인물,', '매니저,', '냉정남,', '순정남,', '무당,', '드라마,액션,', '외유내강,', '판타지,전쟁,멜로

In [ ]:
# 유 : not_nan_df 무: nan_df
train_data = not_nan_df
test_data = nan_df

In [ ]:
# DataFrame 복사
t_data = copy.deepcopy(train_data)

# 워드 임베딩

### Word2Vec

In [ ]:
import re

In [ ]:
from gensim.models import Word2Vec

In [ ]:
z = t_data

In [ ]:
# 각 웹툰의 텍스트 데이터를 담을 리스트 생성
mecab = MeCab()
webtoon_text_data = []


for index, row in z.iterrows():
    title =  str(row['Title']).replace(" ","")
    summary = re.sub('[a-zA-Z0-9]', '', str(row['Summary']))
    category = str(row['Category_Tag'])

    text_data = title + ' ' + ''.join(summary)  + ' ' + category
    webtoon_text_data.append(text_data)
    z.at[index, 'Title'] = title  # 해당 행의 'Title' 값을 수정

In [ ]:
webtoon_text_data[20]

'날것:공작가의하녀로빙의했습니다 남주인공인 줄 알았던 소설 속 최애 캐릭터가, 사실은 서브남이었다?그것도 모자라, 황태자인 남주에게 반란을 일으키고 처형당하는 엔딩?‘…나였으면 그를 행복하게 만들어 줬을 텐데….’서브남앓이를 불쌍하게 본 하늘의 뜻인가?잠에서 깨어나 보니 소설 속 최애의……저택 하녀로 빙의되었다?!온몸을 갉아먹는 저주, 공작 가문의 수치, 사랑에 굶주린 어린 짐승.온갖 비극적 설정을 끌어안은 서브남 ‘리안드로 치릴로 벨라비티’의 일생이 한 엑스트라의 등장으로 변해가기 시작하는데,어째…… 리안드로의 플래그 방향 또한 변하는 것 같은 느낌이…?비틀리는 소설의 전개와 함께 밝혀지는 끔찍한 저주의 진실은?!========================================[표지 및 로고 디자인] 송지혜 로맨스판타지, 힐링물, 치유물, 빙의물'

In [ ]:
# 토큰화된 데이터에 웹툰 제목도 추가
tokenized_data = []
for index, text in enumerate(webtoon_text_data):
    tokens = text.split()
    tokens.extend(tokens[0])  # 웹툰 제목 추가
    tokenized_data.append(tokens)

In [ ]:
# Word2Vec 모델 학습
model = Word2Vec(sentences=tokenized_data, vector_size=5000, window=3, min_count=1, workers=4)

# 완성된 임베딩 매트릭스의 크기 확인
embedding_size = model.wv.vectors.shape
print("Embedding Matrix Size:", embedding_size)


Embedding Matrix Size: (15386, 5000)


In [ ]:
# 학습된 Word2Vec 모델의 전체 어휘 확인
vocab = model.wv.index_to_key
print("Vocabulary:", vocab)

Vocabulary: ['는', '가', '의', '이', '은', '드라마,', '와', '를', ',', '사랑', '다', '판타지,', '한', '1', '을', '수', '멜로/로맨스,', '위해', '과', '라', '사', '아', '된다.', '!', '그', '내', '니', '모험', '에게', '2', '고', '신', '대', '성숙', '자', '어', '도전', '?', '지', '자신의', '하', '한다.', '3', '나', '에', '있는', '액션,', '제자/학생,', '마', '카리스마있음,', '세', '데', '결', '소년영웅,', '부', '드', '로맨스판타지,', '해', '4', '스', '회사원,', '구출', '된', '5', '6', '전', '고아,', '주', '함께', '인', '소', '당', '요', 'SF,', '로', '시', '7', '우', '여', '도', '년', '것을', '다른', '멜로/로맨스,드라마,', '들', '게', '8', '생', '성장물,', '없는', '살', '녀', '혼', '모', '9', '입', '랑', '공', '스릴러,', '만', '날', '짝사랑,', '리', '기', '서', '삼각관계,', '저', '활동', '갑작스런사고,', '잘', '다시', '될', '사랑을', '실', '거지,', '능력녀,', '없', '계', '돌', '렐', '구', '미인,', '쿨', '상', '적', '온', '하는', '콤', '금지된사랑', '알게', '탕', '같은', '현대로맨스,', '두', '되고', '안', '코미디,', '너', '되', '학교', '애', '추구', '황족(왕족),', '기억을', '새로운', '그의', '몰락한귀족,', '모든', '다정남,', '더', '집', '자신을', '알고', '어느', '건', '왕족/귀족,', '판타지,드라마,', '우연히', '0', '능력남,', '대한', '할', '하고', '험', '강제된과업

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

# 유사도 계산을 위한 함수: 코사인 유사도 사용.
def calculate_similarity(title1, title2):
    vec1 = model.wv[title1]
    vec2 = model.wv[title2.replace(" ", "")]
    similarity = cosine_similarity([vec1], [vec2])[0][0]

    # 백분율
    similarity_percentage = (similarity + 1) * 50
    return similarity_percentage

# 사용자 입력 받기
user_input = input("유사한 웹툰을 찾을 키워드를 입력하세요: ")

# 입력을 기반으로 유사한 웹툰 추천
target_webtoon = user_input
recommended_webtoons = train_data.copy()  # 복사본 생성

# 'Similarity' 컬럼 추가하지 않고 바로 출력
similarities = []

print(f'\n{target_webtoon}과(와) 유사한 웹툰 추천:\n')
top_n = 10  # 원하는 상위 N개의 웹툰을 출력하도록 설정
count = 0

for index, row in recommended_webtoons.iterrows():
    similarity = calculate_similarity(target_webtoon, row['Title'])
    similarities.append((row['Title'], similarity))

# 유사도가 높은 순으로 정렬하여 상위 N개 출력
similarities = sorted(similarities, key=lambda x: x[1], reverse=True)[:top_n]
for title, similarity in similarities:
    count += 1
    print(f"{count}. {title} : 유사도: {similarity:.2f}%")


유사한 웹툰을 찾을 키워드를 입력하세요: 일상

일상과(와) 유사한 웹툰 추천:

1. 100마일 : 유사도: 52.71%
2. 리매치 : 유사도: 52.65%
3. 아기 황후님 : 유사도: 52.61%
4. 마법 같은 사랑 : 유사도: 52.60%
5. 마법 같은 사랑 : 유사도: 52.60%
6. 흰 족제비 양에게 죽음을! : 유사도: 52.55%
7. 마술, 좋아하세요? 2 : 유사도: 52.42%
8. 마술, 좋아하세요? 2 : 유사도: 52.42%
9. 외나무다리에 꽃을 뿌려주세요 : 유사도: 52.36%
10. 논현동 장사꾼 : 유사도: 52.35%


In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

# TF-IDF 벡터화
tfidf_vectorizer = TfidfVectorizer()
tfidf_matrix = tfidf_vectorizer.fit_transform(webtoon_text_data)

# 사용자 입력 받기
user_input = input("키워드를 입력하세요: ")

# 입력을 기반으로 유사한 웹툰 추천
target_webtoon = user_input

# 입력한 키워드에 대한 TF-IDF 벡터 계산
target_vector = tfidf_vectorizer.transform([target_webtoon])

# 입력 키워드와의 유사도 계산
cosine_similarities = cosine_similarity(target_vector, tfidf_matrix).flatten()

# 유사도가 높은 순으로 정렬하여 상위 5개 출력
similarities_with_index = list(enumerate(cosine_similarities))
sorted_similarities = sorted(similarities_with_index, key=lambda x: x[1], reverse=True)[:5]

print(f'\n"{target_webtoon}"과(와) 유사한 웹툰 추천:\n')

for i, (index, similarity) in enumerate(sorted_similarities, start=1):
    title = z.iloc[index]['Title']
    print(f"{count}. {title} : 유사도: {similarity:.2f}%")


키워드를 입력하세요: 이세계

"이세계"과(와) 유사한 웹툰 추천:

1. 이세계의황비
2. 악녀를죽여줘
3. 도굴왕
4. 8서클마법사의환생
5. 고백의이유


### 한국어 KoBERT

In [ ]:
from transformers import BertModel, BertTokenizer
from sklearn.metrics.pairwise import cosine_similarity
import torch

In [ ]:
# koBERT 모델과 토크나이저 불러오기
model_name = 'kykim/bert-kor-base'
tokenizer = BertTokenizer.from_pretrained(model_name)
model = BertModel.from_pretrained(model_name)

# 입력 텍스트를 koBERT의 임베딩 벡터로 변환하는 함수
def get_embedding(text):
    inputs = tokenizer(text, return_tensors="pt", truncation=True)
    outputs = model(**inputs)
    return outputs.last_hidden_state.mean(dim=1)


In [ ]:

# 사용자 입력 받기
user_input = input("추천을 원하는 웹툰을 입력하세요: ")

# 입력 텍스트와 각 웹툰 간의 유사도 계산
user_embedding = get_embedding(user_input)
webtoon_embeddings = [get_embedding(text) for text in webtoon_text_data]
similarities = cosine_similarity(user_embedding, webtoon_embeddings).flatten()

# 유사도가 높은 순으로 정렬하여 상위 5개 출력
sorted_similarities = sorted(enumerate(similarities), key=lambda x: x[1], reverse=True)[:5]


print(f'\n"{user_input}"과(와) 유사한 웹툰 추천:\n')

for i, (index, similarity) in enumerate(sorted_similarities, start=1):
    title = f"웹툰{index + 1}"  # 실제 데이터에 맞게 수정 필요
    print(f"{i}. {title} : 유사도: {similarity:.2f}")

추천을 원하는 웹툰을 입력하세요: 로맨스
